include/hyper.deal/numerics/vector_tools.h

```diff
@@ -324,6 +324,106 @@ namespace hyperdeal
+     * and @dof_no_v.
+     */
+    template <int degree,
+              int n_points,
+              int dim_x,
+              int dim_v,
+              typename Number,
+              typename VectorizedArrayType,
+              typename Vector_Out,
+              typename Vector_In>
+    void
+    coordinate_space_integration(
+      const MatrixFree<dim_x, dim_v, Number, VectorizedArrayType> &data,
+      Vector_Out &                                                 dst,
+      const Vector_In &                                            src,
+      const unsigned int                                           dof_no_x,
+      const unsigned int                                           dof_no_v,
+      const unsigned int                                           quad_no_x)
+    {
+      using MF =
+        hyperdeal::MatrixFree<dim_x, dim_v, Number, VectorizedArrayType>;
+
+      using VectorizedArrayTypeX = typename MF::VectorizedArrayTypeX;
+      using VectorizedArrayTypeV = typename MF::VectorizedArrayTypeV;
+
+      AssertDimension(VectorizedArrayTypeV::size(), 1);
+
+      FEEvaluation<dim_x, dim_v, degree, n_points, Number, VectorizedArrayType>
+      phi(const_cast<MatrixFree<dim_x, dim_v, Number, VectorizedArrayType> &>(
+            data),
+          dof_no_x,
+          dof_no_v,
+          0 /*dummy*/,
+          0 /*dummy*/);
+
+      dealii::
+        FEEvaluation<dim_x, degree, n_points, 1, Number, VectorizedArrayTypeX>
+          phi_x(data.get_matrix_free_x(), dof_no_x, quad_no_x);
+      dealii::
+        FEEvaluation<dim_v, degree, n_points, 1, Number, VectorizedArrayTypeV>
+          phi_v(data.get_matrix_free_v(), dof_no_v, 0 /*dummy*/);
+
+      dst = 0.0; // clear destination vector
+
+      data.template cell_loop<Vector_Out, Vector_In>(
+        [&](const auto &,
+            Vector_Out &     dst,
+            const Vector_In &src,
+            const auto       cell) mutable {
+          phi.reinit(cell);
+          phi.read_dof_values(src);
+
+          phi_x.reinit(cell.x);
+          phi_v.reinit(cell.v);
+
+          // get data and scratch
+          const VectorizedArrayType *data_ptr_src = phi.get_data_ptr();
+          VectorizedArrayTypeV *     data_ptr_dst = phi_v.begin_dof_values();
+
+          const unsigned int n_lanes_filled_x =
+            phi.n_vectorization_lanes_filled();
+
+          // loop over all x points and integrate over all v points
+          for (unsigned int qv = 0; qv < phi_v.n_q_points; ++qv)
+            {
+              VectorizedArrayTypeV sum_x = VectorizedArrayTypeV();
+
+              for (unsigned int qx = 0; qx < phi_x.n_q_points; ++qx)
+                for (unsigned int lane_x = 0; lane_x < n_lanes_filled_x;
+                     ++lane_x)
+                  sum_x +=
+                    data_ptr_src[qx + dealii::Utilities::pow(n_points, dim_v) *
+                                        qv][lane_x] *
+                    phi_x.JxW(qx)[lane_x];
```

Comment:
For performance-critical code, I would run the summation over the variable type of `VectorizedArrayType` first and only perform a cross-lane reduction outside the `qx` loop. (We currently hardcode that `VectorizedArrayTypeV` is different from `VectorizedArrayType` anyway, so I also see no benefit in terms of generality by doing the reduction like this.)

Reply:
I have switched the if-statement. But as mentioned in the discussion, this function will not be part of this PR.